### PR TITLE
[AppTheme] Add high contrast properties and events

### PR DESCRIPTION
**Pull request by decrowle**
Adding in High Contrast awareness for React Native.

Adds High contrast event to the AppTheme module. Provides awareness of the 'isHighContrast' Boolean and 'highContrastColors' which are the RGB hex values for the eight main system colors from Windows:
ButtonFace
ButtonText
GrayText
Highlight
HighlightText
Hotlight
Window
WindowText
These are sent via the highContrastChanged event as strings of the form "#FFFFFF", matching conventional RGB color values for React Native. 

Examples are also added to AccessibilityExample.tsx. 
This PR also includes the addition of ThemingExample.tsx which adds an example of usage for the light and dark theme awareness from the AppTheme module.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/2697)

### CODE REVIEW

File: vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp [hunk 1]
```diff
@@ -43,19 +55,52 @@ AppTheme::AppTheme(const std::shared_ptr<IReactInstance>& reactInstance, const s
   });
 }
 
+// Returns the RBG values for the 8 relevant High Contrast elements.
+folly::dynamic AppTheme::getHighContrastColors() {
+  winrt::Windows::UI::Color ButtonFaceColor = m_uiSettings.UIElementColor(winrt::Windows::UI::ViewManagement::UIElementType::ButtonFace);
```
Comment (marlenecota) [inline vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp]: >winrt::Windows::UI::ViewManagement [](start = 74, length = 34)

nit: Probably personal preference, but it's a little too long for me. You could add Windows::UI::ViewManagement to namespace winrt at the too and just do "winrt::UIElementType::ButtonFace"

File: vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp [hunk 2]
```diff
@@ -43,19 +55,52 @@ AppTheme::AppTheme(const std::shared_ptr<IReactInstance>& reactInstance, const s
...
+  winrt::Windows::UI::Color ButtonTextColor = m_uiSettings.UIElementColor(winrt::Windows::UI::ViewManagement::UIElementType::ButtonText);
+  winrt::Windows::UI::Color GrayTextColor = m_uiSettings.UIElementColor(winrt::Windows::UI::ViewManagement::UIElementType::GrayText);
+  winrt::Windows::UI::Color HighlightColor = m_uiSettings.UIElementColor(winrt::Windows::UI::ViewManagement::UIElementType::Highlight);
+  winrt::Windows::UI::Color HighlightTextColor = m_uiSettings.UIElementColor(winrt::Windows::UI::ViewManagement::UIElementType::HighlightText);
+  winrt::Windows::UI::Color HotlightColor = m_uiSettings.UIElementColor(winrt::Windows::UI::ViewManagement::UIElementType::Hotlight);
+  winrt::Windows::UI::Color WindowColor = m_uiSettings.UIElementColor(winrt::Windows::UI::ViewManagement::UIElementType::Window);
+  winrt::Windows::UI::Color WindowTextColor = m_uiSettings.UIElementColor(winrt::Windows::UI::ViewManagement::UIElementType::WindowText);
+  folly::dynamic rbgValues = folly::dynamic::object("ButtonFaceColor", formatRGB(ButtonFaceColor))
```
Comment (marlenecota) [inline vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp]: >folly::dynamic rbgValues = folly::dynamic::object("ButtonFaceColor", formatRGB(ButtonFaceColor)) [](start = 2, length = 96)

nit: Add blank line separating the two

File: vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
 #include "pch.h"
 #include "AppThemeModuleUwp.h"
+#include <iomanip>
```
Comment (marlenecota) [inline vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp]: What's this for?

File: vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp [hunk 4]
```diff
@@ -43,19 +55,52 @@ AppTheme::AppTheme(const std::shared_ptr<IReactInstance>& reactInstance, const s
   });
 }
 
+// Returns the RBG values for the 8 relevant High Contrast elements.
+folly::dynamic AppTheme::getHighContrastColors() {
```
Comment (marlenecota) [inline vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp]: Keep the function order the same as in the header file

File: vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp (hunk 3 above)
Comment (decrowle) [reply vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp]: The string manipulation functions for formatting the RGB values into the hex strings.

File: vnext/ReactWindowsCore/Modules/AppThemeModule.cpp [hunk 5]
```diff
@@ -19,6 +19,16 @@ const std::string AppTheme::getCurrentTheme()
...
 }
 
+bool AppTheme::getIsHighContrast()
+{
+  return false;
+}
+
+folly::dynamic AppTheme::getHighContrastColors()
+{
+  return folly::dynamic::object("highContrastColors", "None");
```
Comment (marlenecota) [inline vnext/ReactWindowsCore/Modules/AppThemeModule.cpp]: >folly::dynamic::object("highContrastColors", "None"); [](start = 9, length = 53)

return {} instead.

File: vnext/src/Libraries/AppTheme/AppTheme.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  ButtonTextColor: string;
+  GrayTextColor: string;
+  HighlightColor: string;
+  HighlightTextColor: string;
+  HotlightColor: string;
+  WindowColor: string;
+  WindowTextColor: string;
+}
+
+export interface IAppThemeChangedEvent {
```
Comment (marlenecota) [inline vnext/src/Libraries/AppTheme/AppTheme.ts]: >IAppThemeChangedEvent [](start = 17, length = 21)

Remove currentTheme and change to IHighContrastChangeEvent.
Add IAppThemeChangeEvent with only currentTheme in it.

File: vnext/src/Libraries/AppTheme/AppTheme.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  isHighContrast: boolean;
+  highContrastColors: IColorValues;
+  currentTheme: string;
+}
```
Comment (marlenecota) [inline vnext/src/Libraries/AppTheme/AppTheme.ts]: This should be a different file called AppThemeTypes or something (they're not really props)

File: vnext/src/Libraries/Modules/AppTheme/AppTheme.ts [hunk 8]
```diff
@@ -1,14 +0,0 @@
...
-import { NativeEventEmitter } from 'react-native';
-
-class AppThemeModule extends NativeEventEmitter  {
-  get currentTheme(): string {
-    return '';
-  }
-}
-
-export const AppTheme = new AppThemeModule();
-export default AppTheme;
```
Comment (marlenecota) [inline vnext/src/Libraries/Modules/AppTheme/AppTheme.ts]: Add this file back and isHighContrast.

**Comment by marlenecota** [comment]
> class AppThemeModule extends NativeEventEmitter  {

add isHighContrast

---
Refers to: vnext/src/Libraries/AppTheme/__mocks__/AppTheme.uwp.ts:6 in cbd5625. [](commit_id = cbd56253bd658bbc4bb1c3eab4aa82ac7d4b4fa8, deletion_comment = False)

File: vnext/src/RNTester/AccessibilityExample.tsx [hunk 9]
```diff
@@ -56,61 +115,6 @@ class TouchableExamples extends React.Component<{}, any> {
   }
 }
 
-class AccessibilityStateExamples extends React.Component {
```
Comment (marlenecota) [inline vnext/src/RNTester/AccessibilityExample.tsx]: Did you mean to get rid of this example?

File: vnext/src/RNTester/AccessibilityExample.tsx [hunk 10]
```diff
@@ -4,21 +4,80 @@
...
+
+  public render() {
+    return (
+      <View>
+        <Text>The following has HighContrast Event awareness:</Text>
+        <View
+          style={{width: 150, height: 50, backgroundColor: currentBackgroundColor[2]}}
+          accessibilityLabel="A blue box"
+          accessibilityHint="A hint for the blue box.">
+        <Text>isHighContrast: {this.state.isHighContrast ? 'true' : 'false'}</Text>
```
Comment (marlenecota) [inline vnext/src/RNTester/AccessibilityExample.tsx]: ><Text>isHighContrast: {this.state.isHighContrast ? 'true' : 'false'}</Text> [](start = 8, length = 75)

One more tab

File: vnext/src/RNTester/AccessibilityExample.tsx [hunk 11]
```diff
@@ -4,21 +4,80 @@
 /* tslint:disable */
 
 import React = require('react');
-import { FlatList, Text, TouchableHighlight, View } from 'react-native';
+import { Text, TouchableHighlight, View } from 'react-native';
+import { AppTheme } from '../../src/index.uwp';
+import { IAppThemeChangedEvent } from 'src/Libraries/AppTheme/AppTheme';
+
+var currentBackgroundColor = ['blue', 'red', '#E6E6E6', '#7fff00'];
```
Comment (marlenecota) [inline vnext/src/RNTester/AccessibilityExample.tsx]: Doesn't look like you need this. Inline the style or use a stylesheet.

File: vnext/src/RNTester/AccessibilityExample.tsx [hunk 12]
```diff
@@ -4,21 +4,80 @@
...
+        <Text>isHighContrast: {this.state.isHighContrast ? 'true' : 'false'}</Text>
+        </View>
+        <Text>ButtonFaceColor value: {this.state.highContrastColorValues.ButtonFaceColor}</Text>
+        <Text>ButtonTextColor value: {this.state.highContrastColorValues.ButtonTextColor}</Text>
+        <Text>GrayTextColor value: {this.state.highContrastColorValues.GrayTextColor}</Text>
+        <Text>HighlightColor value: {this.state.highContrastColorValues.HighlightColor}</Text>
+        <Text>HighlightTextColor value: {this.state.highContrastColorValues.HighlightTextColor}</Text>
+        <Text>HotlightColor value: {this.state.highContrastColorValues.HotlightColor}</Text>
+        <Text>WindowColor value: {this.state.highContrastColorValues.WindowColor}</Text>
+        <Text>WindowTextColor value: {this.state.highContrastColorValues.WindowTextColor}</Text>
```
Comment (marlenecota) [inline vnext/src/RNTester/AccessibilityExample.tsx]: nit: A more interesting example would be doing a simple version of what's shown in PC settings. If highcontrast is disabled, then grey out the view. If it is enabled, then show the colors in action on buttons/text.

File: vnext/src/RNTester/ThemingExample.uwp.tsx [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+  };
+
+  _onPress = () => {
+  }
+
+  public render() {
+    return (
+      <View>
+        <Text style={{color: 'red'}}>currentTheme: {this.state.currentTheme}</Text>
+        <Button onPress={ this._onPress } title={this.state.currentTheme === 'dark' ? 'Dark' : 'Light'} color={this.state.currentTheme === 'dark' ? 'grey' : 'orange'}></Button>
```
Comment (marlenecota) [inline vnext/src/RNTester/ThemingExample.uwp.tsx]: >=== 'dark' ? 'Dark' : 'Light' [](start = 73, length = 29)

nit: you can remove this and just have the label be lowercase.

Comment (marlenecota) [inline vnext/src/RNTester/ThemingExample.uwp.tsx]: >onPress={ this._onPress }  [](start = 16, length = 26)

This doesn't do anything. Does a button require an onPress handler? If not, remove it.

File: vnext/src/index.ts [hunk 14]
```diff
@@ -11,4 +11,4 @@ export * from './Libraries/Components/Keyboard/KeyboardExt';
 export * from './Libraries/Components/Keyboard/KeyboardExtProps';
 export * from './Libraries/Components/View/ViewWindowsProps';
 export * from './Libraries/Components/View/ViewWindows';
-export * from './Libraries/Modules/AppTheme/AppTheme';
+export * from './Libraries/AppTheme/AppTheme';
```
Comment (marlenecota) [inline vnext/src/index.ts]: Also export AppThemeTypes

File: vnext/src/Libraries/AppTheme/AppThemeTypes.ts [hunk 15]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+export interface IColorValues  {
```
Comment (marlenecota) [inline vnext/src/Libraries/AppTheme/AppThemeTypes.ts]: >IColorValues [](start = 17, length = 12)

Rename to IHighContrastColors

**Comment by marlenecota** [comment]
>   public currentTheme = '';

add isHighContrast and currentHighContrastColorValues to shim

---
Refers to: vnext/src/Libraries/AppTheme/AppTheme.uwp.ts:51 in 70487b7. [](commit_id = 70487b7cdffceccc24e8af9ce210b5b0d141fafb, deletion_comment = False)

File: vnext/src/Libraries/AppTheme/AppTheme.uwp.ts [hunk 16]
```diff
@@ -24,6 +34,14 @@ class AppThemeModule extends NativeEventEmitter  {
   get currentTheme(): string {
     return this._currentTheme;
   }
+
+  get isHighContrast(): boolean {
+    return this._isHighContrast;
+  }
+
+  get currentHighContrastColorValues(): IColorValues {
```
Comment (marlenecota) [inline vnext/src/Libraries/AppTheme/AppTheme.uwp.ts]: >currentHighContrastColorValues [](start = 6, length = 30)

consider renaming to highContrastColors or currentHighContrastColors

File: vnext/src/index.uwp.ts [hunk 17]
```diff
@@ -11,4 +11,4 @@ export * from './Libraries/Components/Keyboard/KeyboardExt.uwp';
 export * from './Libraries/Components/Keyboard/KeyboardExtProps';
 export * from './Libraries/Components/View/ViewWindowsProps';
 export * from './Libraries/Components/View/ViewWindows.uwp';
-export * from './Libraries/Modules/AppTheme/AppTheme.uwp';
\ No newline at end of file
+export * from './Libraries/AppTheme/AppTheme.uwp';
```
Comment (marlenecota) [inline vnext/src/index.uwp.ts]: Not sure if needed, but probably also export AppThemeTypes here as well

File: vnext/src/Libraries/AppTheme/__mocks__/AppTheme.uwp.ts [hunk 18]
```diff
@@ -2,11 +2,20 @@
 'use strict';
```
Comment (marlenecota) [inline vnext/src/Libraries/AppTheme/__mocks__/AppTheme.uwp.ts]: Spoke with @ahimberg , the shim seems to be working, so we can remove this file/folder.

File: vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp [hunk 19]
```diff
@@ -50,12 +63,46 @@ const std::string AppTheme::getCurrentTheme()
...
+  return rbgValues;
+}
+
+std::string AppTheme::formatRGB(winrt::Windows::UI::Color ElementColor) {
+  std::stringstream RGBString;
+  RGBString << "#" << std::setfill('0') << std::setw(2) << std::hex << (int)ElementColor.R 
+            << std::setfill('0') << std::setw(2) << std::hex << (int)ElementColor.G 
+            << std::setfill('0') << std::setw(2) << std::hex << (int)ElementColor.B;
+  return RGBString.str();
+}
```
Comment (ahimberg) [inline vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp]: Our version of this code does this:
 char colorChars[10];
 sprintf_s(colorChars, "#%02x%02x%02x%02x", color.R, color.G, color.B, color.A);
though whichever you find more readable is probably fine.

File: vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp (hunk 3 above)
Comment (ahimberg) [inline vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp]: you probably don't need iomanip now if youre using sprintf_s?

File: vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp [hunk 20]
```diff
@@ -50,12 +62,44 @@ const std::string AppTheme::getCurrentTheme()
   return m_currentTheme == winrt::ApplicationTheme::Light ? AppTheme::light : AppTheme::dark;
 }
 
+bool AppTheme::getIsHighContrast()
+{
+  return m_accessibilitySettings.HighContrast();;
```
Comment (kmelmon) [inline vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp]: >; [](start = 48, length = 1)

nit:  extra semicolon

Comment (decrowle) [reply vnext/ReactUWP/Modules/AppThemeModuleUwp.cpp]: Shoot, merged before the page fully loaded. Will fix this on next PR.

**Pull request merged**